### DISCUSSION TITLE
Tracking outputs in HT3D

### DISCUSSION THREAD

**Discussion by Vojta-Salek** [firemodels/fds#13246] [Q&A]
Hello FDS community,
Is there a way to cleverly record any quantity through multiple stacked obstacles in HT3D mode? Basically something like slice file for gas phase. I would also like to keep track of for example the average density of a selected OBST in time (assuming, for example, that it decomposes). However, the spatial statistic is not available for 'SOLID DENSITY' quantity. It is possible to display many quantities in certain depth, but this approach may not be sufficient in some cases.

I have also created a very simple test case in FDS 6.9.1 to track some basic quantities in the easiest possible geometry - two one-cell thick OBSTs on top of each other (the FDS code is below). Each OBST has its own MATL (MATL_1 and MATL_2). From the top surface, quantities are tracked via DEVCs in various depths. I was really surprised by most of the obtained results, apart from thickness which behaves predictably. For example:

1) Mass fraction of MATL_1 is always 1 independent of depth. Even in a space (depth) where only the second material is supposed to be or outside the defined OBSTs. Mass fraction of MATL_2 is always 0.
2) Specific heat capacity acts similarly but for MATL_2, 'Inf' value is reported.
3) Conductivity is always same and equal to conductivity of MATL_1 (independent of depth or MATL_ID).
4) Density is always same and equal to density of MATL_2 (independent of depth or MATL_ID).

The outputs seem random, I would expect it to switch between the corresponding quantity values based on the selected combination of depth and MATL_ID. Do you have any idea where I'm going wrong? (I am aware that some of DEVCs do not really make sense, I have created them for testing purposes before dealing with some more advanced setups.)

Thank you for any comments in advance!

Vojtech Salek

------------------------------
The FDS code:
&HEAD CHID='HT3D devices', /
&MESH ID='gas', IJK=5,5,5, XB=0.0, 0.05, 0.0, 0.05, 0.0, 0.05 /
&TIME T_END=60. / 
&DUMP DT_DEVC = 1. /
     
&MATL ID                      = 'MATL_1',
      SPECIFIC_HEAT        = 1.0,
      CONDUCTIVITY       = 0.50,
      DENSITY                  = 1000 /

&MATL ID                      = 'MATL_2',
      SPECIFIC_HEAT        = 2.0,
      CONDUCTIVITY       = 1.0,
      DENSITY                  = 2000. / 

&SURF ID                      = 'SURF_1',
      COLOR                    = 'SKY BLUE',         
      HT3D                      = T /

&SURF ID                      = 'SURF_2',
      COLOR                    = 'WHITE',         
      HT3D                      = T /

&SURF ID                     = 'CONE',
      NET_HEAT_FLUX     = 300.
      COLOR                   = 'RED'/

&OBST XB=0.02,0.03,0.02,0.03,0.02,0.03, MATL_ID='MATL_1', SURF_ID= 'SURF_1' /   
&OBST XB=0.02,0.03,0.02,0.03,0.03,0.04, MATL_ID='MATL_2', SURF_ID= 'SURF_2' /   

&DEVC ID='Thickness', XYZ=0.025,0.025,0.04, IOR=3, QUANTITY='WALL THICKNESS' /

&DEVC ID='MATL_1_MF_0.005', XYZ=0.025,0.025,0.04, IOR=3, QUANTITY='SOLID MASS FRACTION', MATL_ID='MATL_1', DEPTH = 0.005 /
&DEVC ID='MATL_1_MF_0.015', XYZ=0.025,0.025,0.04, IOR=3, QUANTITY='SOLID MASS FRACTION', MATL_ID='MATL_1', DEPTH = 0.015 /
&DEVC ID='MATL_1_MF_0.025', XYZ=0.025,0.025,0.04, IOR=3, QUANTITY='SOLID MASS FRACTION', MATL_ID='MATL_1', DEPTH = 0.025 /
&DEVC ID='MATL_2_MF_0.005', XYZ=0.025,0.025,0.04, IOR=3, QUANTITY='SOLID MASS FRACTION', MATL_ID='MATL_2', DEPTH = 0.005 /
&DEVC ID='MATL_2_MF_0.015', XYZ=0.025,0.025,0.04, IOR=3, QUANTITY='SOLID MASS FRACTION', MATL_ID='MATL_2', DEPTH = 0.015 /

&DEVC ID='cp_0.005', XYZ=0.025,0.025,0.04, IOR=3, QUANTITY='SOLID SPECIFIC HEAT', MATL_ID='', DEPTH = 0.005 /
&DEVC ID='cp_0.015', XYZ=0.025,0.025,0.04, IOR=3, QUANTITY='SOLID SPECIFIC HEAT', MATL_ID='', DEPTH = 0.015 /
&DEVC ID='MATL_1_cp_0.005', XYZ=0.025,0.025,0.04, IOR=3, QUANTITY='SOLID SPECIFIC HEAT', MATL_ID='MATL_1', DEPTH = 0.005 /
&DEVC ID='MATL_1_cp_0.015', XYZ=0.025,0.025,0.04, IOR=3, QUANTITY='SOLID SPECIFIC HEAT', MATL_ID='MATL_1', DEPTH = 0.015 /
&DEVC ID='MATL_2_cp_0.005', XYZ=0.025,0.025,0.04, IOR=3, QUANTITY='SOLID SPECIFIC HEAT', MATL_ID='MATL_2', DEPTH = 0.005 /
&DEVC ID='MATL_2_cp_0.015', XYZ=0.025,0.025,0.04, IOR=3, QUANTITY='SOLID SPECIFIC HEAT', MATL_ID='MATL_2', DEPTH = 0.015 /

&DEVC ID='k_0.005', XYZ=0.025,0.025,0.04, IOR=3, QUANTITY='SOLID CONDUCTIVITY', MATL_ID='', DEPTH = 0.005 /
&DEVC ID='k_0.015', XYZ=0.025,0.025,0.04, IOR=3, QUANTITY='SOLID CONDUCTIVITY', MATL_ID='', DEPTH = 0.015 /
&DEVC ID='MATL_1_k_0.005', XYZ=0.025,0.025,0.04, IOR=3, QUANTITY='SOLID CONDUCTIVITY', MATL_ID='MATL_1', DEPTH = 0.005 /
&DEVC ID='MATL_1_k_0.015', XYZ=0.025,0.025,0.04, IOR=3, QUANTITY='SOLID CONDUCTIVITY', MATL_ID='MATL_1', DEPTH = 0.015 /
&DEVC ID='MATL_2_k_0.005', XYZ=0.025,0.025,0.04, IOR=3, QUANTITY='SOLID CONDUCTIVITY', MATL_ID='MATL_2', DEPTH = 0.005 /
&DEVC ID='MATL_2_k_0.015', XYZ=0.025,0.025,0.04, IOR=3, QUANTITY='SOLID CONDUCTIVITY', MATL_ID='MATL_2', DEPTH = 0.015 /

&DEVC ID='MATL_1_rho_0.005', XYZ=0.025,0.025,0.04, IOR=3, QUANTITY='SOLID DENSITY', MATL_ID='MATL_1', DEPTH = 0.005 /
&DEVC ID='MATL_2_rho_0.015', XYZ=0.025,0.025,0.04, IOR=3, QUANTITY='SOLID DENSITY', MATL_ID='MATL_2', DEPTH = 0.015 /
&DEVC ID='rho_0.005', XYZ=0.025,0.025,0.04, IOR=3, QUANTITY='SOLID DENSITY', MATL_ID='', DEPTH = 0.005 /
&DEVC ID='rho_0.015', XYZ=0.025,0.025,0.04, IOR=3, QUANTITY='SOLID DENSITY', MATL_ID='', DEPTH = 0.015 /

&DEVC ID='T_0.005', XYZ=0.025,0.025,0.04, IOR=3, QUANTITY='INSIDE WALL TEMPERATURE', DEPTH = 0.005 /
&DEVC ID='T_0.015', XYZ=0.025,0.025,0.04, IOR=3, QUANTITY='INSIDE WALL TEMPERATURE', DEPTH = 0.015 /

&VENT XB=0.01,0.04,0.01,0.04,0.0, 0.0, IOR=3, SURF_ID='CONE' /
&VENT MB='XMIN', SURF_ID = 'OPEN' /
&VENT MB='XMAX', SURF_ID = 'OPEN' /
&VENT MB='YMIN', SURF_ID = 'OPEN' /
&VENT MB='YMAX', SURF_ID = 'OPEN' /
&VENT MB='ZMAX', SURF_ID = 'OPEN' /

&TAIL /
[HT3D devices_devc.csv](https://github.com/user-attachments/files/16421921/HT3D.devices_devc.csv)

**Comment by mcgratta** [comment]
You can use a `'TEMPERATURE'` slice to view inner temperature, at least at the resolution of the gas phase grid. In Smokeview make the obstruction outline only and then Show/Hide --> Slice --> Gas/Solid. I'll look at the other issues.

**Reply by Vojta-Salek** [reply]
Thank you very much, did not know about this Smokeview feature. Does it also work for other slice quantities than 'TEMPERATURE'? 

Just for other possible readers: After hiding the obstructions (in case there is multiple rows of objects) or make them outline only (single row), I had to Load/unload --> Slice --> Settings.. --> show slice region (check 'solid') & draw slice using (check 'solid colors').

**Reply by mcgratta** [reply]
It only works for `'TEMPERATURE'`.

**Comment by mcgratta** [comment]
I have found several bugs related to this type of output. Do not use this feature until I fix and post a new version.

**Reply by Vojta-Salek** [reply]
Understood. Thank you for looking into it.

**Comment by mcgratta** [comment]
[New version is here](https://github.com/firemodels/test_bundles/releases/tag/FDS_TEST)
